mithril-common/src/chain_observer/pallas_observer.rs

```diff
@@ -0,0 +1,247 @@
+use anyhow::{anyhow, Context};
+use async_trait::async_trait;
+use pallas_network::facades::NodeClient;
+use pallas_network::miniprotocols::localstate::{queries_v16, Client};
+use std::path::{Path, PathBuf};
+
+use crate::chain_observer::interface::*;
+use crate::chain_observer::{ChainAddress, TxDatum};
+use crate::crypto_helper::{KESPeriod, OpCert};
+use crate::entities::StakeDistribution;
+use crate::CardanoNetwork;
+use crate::{entities::Epoch, StdResult};
+
+use super::CardanoCliChainObserver;
+
+/// A runner that uses Pallas library to interact with a Cardano node using N2C Ouroboros mini-protocols
+pub struct PallasChainObserver {
+    socket: PathBuf,
+    network: CardanoNetwork,
+    fallback: super::cli_observer::CardanoCliChainObserver,
+}
+
+impl From<anyhow::Error> for ChainObserverError {
+    fn from(err: anyhow::Error) -> Self {
+        ChainObserverError::General(err)
+    }
+}
+
+impl PallasChainObserver {
+    /// Creates a new PallasObserver while accepting a fallback CliRunner
+    pub fn new(socket: &Path, network: CardanoNetwork, fallback: CardanoCliChainObserver) -> Self {
+        Self {
+            socket: socket.to_owned(),
+            network,
+            fallback,
+        }
+    }
+
+    /// Creates and returns a new `NodeClient` connected to the specified socket.
+    async fn new_client(&self) -> StdResult<NodeClient> {
+        let magic = self.network.code();
+        let client = NodeClient::connect(&self.socket, magic).await?;
+
+        Ok(client)
+    }
+
+    /// Returns a reference to the fallback `CardanoCliChainObserver` instance.
+    fn get_fallback(&self) -> StdResult<&CardanoCliChainObserver> {
```

Comment:
This function could simply return a `&CardanoCliChainObserver`:
```suggestion
    fn get_fallback(&self) -> &CardanoCliChainObserver {
```